README.md
# T-Rex-Runner using CNN-keras


This project is about teaching the T-Rex to dodge obstacles in 3 stages. It is a CNN model that predicts objects and keyboard actions.

* Getting the data
* Training the CNN model
* Testing the trained CNN model

### Getting the data

We need below libraries to collect the data.
```angular2html
import keyboard
import uuid
import time
from PIL import Image
from mss import mss
```

* **keyboard** library is used to save our action like "up", "down" and "right".
* Game screen can be recorded using **mss** library. 
* When all the lines are executed, we will switch the game screen and the model that is trained with the data we got, will predict action according to images. Also, this library helps to cut off some areas in the screen.
* A function will be created which will be used for recording the screen.
* An exit function will be defined and whenever we want to stop collecting data, then we can press `esc` which calls exit function and stops collecting the data.

### Training a CNN model

A CNN model will be trained with the data we got. For that we need below libraries.

```angular2html
import glob
import os
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPooling2D
from PIL import Image
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.model_selection import train_test_split
```

* The model will predict keyboard actions according to images. 
* If the model recognizes a cactus, the action will be "UP", or if the model recognizes a bird, the action will be "DOWN" or "UP".
* Labels(UP, DOWN, RIGHT) will be determined and before training, recorded images will be resized and normalized.
* Data will be split into training and testing data.
* Firstly Label Encoding and then One Hot Encoding will be applied on labels by which a binary value will be generated for each label.
* A CNN will be created in the next step which we will train and test.


### Testing the trained CNN model

* We will test our model in the game as real-time.
* We will import below libraries.

```angular2html
from keras.models import model_from_json
import numpy as np
from PIL import Image
import keyboard
import time
import os
from mss import mss
```

* We will upload our Trained CNN Model and make it choose in the list according to output of the model.
* Resize and normalization will be done and turned into array and reshaped it. Because the input is known for model. 
* The np.argmax() returns the indices of the maximum values along an axis. We will use it due to find the label maximum probability.
* If the result is 0 it means DOWN, using keyboard library, model will press DOWN. 
* If the result is 2 it means UP, again using keyboard library model will press UP.





CHANGELOG.md
# CHANGELOG

### ADDED

* Created an empty markdown file `README.md`.
* Created a directory named `docs`.
* Created a markdown file `PROPOSAL.md` in `docs` directory for writing project proposal.
* Created a markdown file `DESIGN.md` in `docs` directory for writing project design.
* Created a directory called `src` for code development.
* Created a file `trex_getdata.py` in `src` directory for developing the code for collecting the data.
* Created a directory called `data` to store the images got when run the file `trex_getdata.py`.
* Created a file `trex_train.py` in `src` directory for developing code to train the CNN model using keras.
* Created a file `trex_play.py` in `src` directory for developing code to make the trex play the game.
* Generated `requirements.txt` file using the command: `pip freeze > requirements.txt`.
* Created a markdown file `HOWTO.md` to write how to clone and run the project.
* Created a markdown file `CHANGELOG.md` to keep note of all changes made to project.

### UPDATED

* Updated `PROPOSAL.md` with project proposal.
* Updated introduction and theoretical background for project in `DESIGN.md`.
* Updated the approach for developing the project in `DESIGN.md`.
* Updated the working example and ethical considerations for the project in `DESIGN.md`.
* Updated code in `trex_getdata.py`.
* Updated code in `trex_train.py`.
* Updated code in `trex_play.py`.
* Updated project description in `README.md`.
* Updated `HOWTO.md`.
* Updated `CHANGELOG.md`.

### FIXED

* Fixed pylint errors in `trex_getdata.py`.
* Fixed pylint errors in `trex_train.py`.
* Updated the path to the images generated in `trex_train.py` to fix FileNotFound error.
* Changed the entropy type to 'sparse_categorical_crossentropy' from 'categorical_crossentropy' in `trex_train.py` to fix the value error.
* Fixed pylint errors in `trex_play.py`.


### REMOVED

* Removed .idea folder from remote repository


src/getdata.py
import keyboard
import os
import uuid
from mss import mss
from PIL import Image
import time
import sys

mon = {'top': 350, 'left': 625, 'width': 650, 'height': 130}
sct = mss()

counter = 0


def record_screen(record_id1, key):
    """Records screen in the game."""
    global counter
    counter += 1
    print("{}: {}".format(key, counter))
    data = sct.grab(mon)
    im = Image.frombytes('RGB', data.size, data.rgb)
    im.save("../data/{}_{}_{}.png".format(key, record_id1, counter))


is_exit = False


def exit1():
    """Stops collection of data."""
    global is_exit
    is_exit = True


keyboard.add_hotkey("esc", exit1)

record_id = uuid.uuid4()
while True:

    if is_exit:
        break

    try:
        if keyboard.is_pressed(keyboard.KEY_UP):
            record_screen(record_id, "up")
            time.sleep(0.1)
        elif keyboard.is_pressed(keyboard.KEY_DOWN):
            record_screen(record_id, "down")
            time.sleep(0.1)
        elif keyboard.is_pressed("right"):
            record_screen(record_id, "right")
            time.sleep(0.2)
    except RuntimeError:
        continue


src/trex_play.py
"""
File: trex_play.py
Purpose: Trex played by AI for team project
Author: Vyshnavi Mulakalapalli
Created: Dec 05, 2022
Updated: Dec 05, 2022
"""

import os
import time
from keras.models import model_from_json
# import glob
import numpy as np
from PIL import Image
import keyboard
from mss import mss


mon = {'top': 350, 'left': 625, 'width': 650, 'height': 130}
sct = mss()

# Force to use only CPU
os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"   # see issue #152
os.environ["CUDA_VISIBLE_DEVICES"] = ""

WIDTH = 200
HEIGHT = 40

# Load Model
# model = model_from_json(open("model.json", "r", encoding="utf-8").read())
with open("model.json", "r", encoding="utf-8") as model1:
    model = model_from_json(model1.read())
model.load_weights("trex_weight.h5")
print(model.summary())

# Down = 0, Right = 1, UP = 2
labels = ["Down", "Right", "Up"]


framerate_time = time.time()
COUNTER = 0
i = 0
CURRENT_FRAMERATE = 0
DELAY = 0.04
COUNT_TIME = 0
LAST_RESULT = 1

while True:

    time_start = time.time()
    img = sct.grab(mon)
    im = Image.frombytes('RGB', img.size, img.rgb)
    im = np.array(im.convert("L").resize((WIDTH, HEIGHT)))
    im = im / 255

    time_grab = time.time() - time_start
    time_start = time.time()

    X = np.array([im])
    X = X.reshape(X.shape[0], WIDTH, HEIGHT, 1)
    r = model.predict(X)
    result = np.argmax(r)

    time_predict = time.time() - time_start

    if (result == 0) and (LAST_RESULT != 0):
        # Hold down button
        keyboard.press(keyboard.KEY_DOWN)
    elif result == 2:
        if LAST_RESULT == 0:
            # Release down button
            keyboard.release(keyboard.KEY_DOWN)
        keyboard.press_and_release(keyboard.KEY_UP)

    last_result = result

    COUNTER += 1
    if (time.time() - framerate_time) > 1:
        CURRENT_FRAMERATE = COUNTER / (time.time() - framerate_time)
        COUNTER = 0
        framerate_time = time.time()
        DELAY -= 0.0001
        DELAY = max(DELAY, 0)

    os.system('clear')
    print(f"Down: {r[0][0]:3.2%} \nRight: {r[0][1]:3.2%} \n"
          f"Up: {r[0][2]:3.2%} \n", end="\r")
    print("=========================")
    print(f"Frame Count: {i}\nCurrent Frame Rate: {CURRENT_FRAMERATE:.2f}\n"
          f"Grab Screen: {time_grab:.5f}\nPredict: {time_predict:.5f}")
    print("=========================")
    print(f"Delay: {DELAY:.5f}")
    i += 1
    time.sleep(DELAY)


src/trex_train.py
"""
File: trex_train.py
Purpose: train the T-rex dino AI system using CNN
Author: Prasanna
Created: Dec 05,200
Updated: Dec 05,2022.
"""
import glob
import os
import random
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from PIL import Image
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split


def onehot_labels(values):
    """
    Label Encoding -> One Hot Encoding

    convert labels to binary values
    firstly labels will be numeric for example;
    UP --> 0
    DOWN --> 1
    RIGHT --> 2
    After Label Encoding will apply One Hot Encoding
    Thus, labels will be Binary Value, for example,
    0 --> 000
    1 --> 010
    2 -->001
    """
    label_encoder = LabelEncoder()
    integer_encoded = label_encoder.fit_transform(values)
    onehot_encoder = OneHotEncoder(sparse=False)
    integer_encoded = integer_encoded.reshape(len(integer_encoded), 1)
    onehot_encoded = onehot_encoder.fit_transform(integer_encoded)
    return onehot_encoded


imgs = glob.glob("../data/*.png")

WIDTH = 200
HEIGHT = 40

X = []
Y = []
for data in imgs:
    filename = os.path.basename(data)
    label = filename.split("_")[0]
    im = np.array(Image.open(data).convert("L").resize((WIDTH, HEIGHT)))
    im = im / 255
    X.append(im)
    Y.append(label)

X = np.array(X)
X = X.reshape(X.shape[0], WIDTH, HEIGHT, 1)
Y = onehot_labels(Y)

train_X, test_X, train_y, test_y = train_test_split(X, Y, test_size=0.25,
                                                    random_state=random.randint
                                                    (112, 1112))

model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=(WIDTH, HEIGHT, 1)))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.4))
model.add(Dense(3, activation='softmax'))

if os.path.exists("./trex_weight.h5"):
    model.load_weights('trex_weight.h5')
    print("Load weight file")

print(model.summary())

model.compile(loss='sparse_categorical_crossentropy', optimizer='SGD',
              metrics=['accuracy'])

model.fit(train_X, train_y, epochs=20, batch_size=64)

scores = model.evaluate(train_X, train_y)
print("\n%s: %.2f%%" % (model.metrics_names[1], scores[1] * 100))

scores = model.evaluate(test_X, test_y)
print("\nTEST %s: %.2f%%" % (model.metrics_names[1], scores[1] * 100))

print(model.evaluate(X, Y))
with open("model.json", "w", encoding='utf8') as file:
    file.write(model.to_json())
# open("model.json", "w", encoding='utf8').write(model.to_json())
model.save_weights('trex_weight.h5')


src/trex_getdata.py
"""
File: trex_getdata.py
Purpose: Collects data for team project
Author: Vyshnavi Mulakalapalli
Created: Dec 05, 2022
Updated: Dec 13, 2022
"""
import uuid
import time
import keyboard
# import os
from mss import mss
from PIL import Image
# import sys

mon = {'top': 350, 'left': 625, 'width': 650, 'height': 130}
sct = mss()

COUNTER = 0


def record_screen(record_id1, key):
    """
    Records screens in the game and saves images into data folder.

    record_id1: uuid generated when a key is pressed.
    key: the keyboard action, for example UP, RIGHT or DOWN.
    """
    global COUNTER
    COUNTER += 1
    print(f"{key}: {COUNTER}")
    data = sct.grab(mon)
    im_dt = Image.frombytes('RGB', data.size, data.rgb)
    im_dt.save(f"../data/{key}_{record_id1}_{COUNTER}.png")


IS_EXIT = False


def exit1():
    """Stops collection of data."""
    global IS_EXIT
    IS_EXIT = True


keyboard.add_hotkey("esc", exit1)

record_id = uuid.uuid4()
while True:

    if IS_EXIT:
        break

    try:
        if keyboard.is_pressed(keyboard.KEY_UP):
            record_screen(record_id, "up")
            time.sleep(0.1)
        elif keyboard.is_pressed(keyboard.KEY_DOWN):
            record_screen(record_id, "down")
            time.sleep(0.1)
        elif keyboard.is_pressed("right"):
            record_screen(record_id, "right")
            time.sleep(0.2)
    except RuntimeError:
        continue
